modules/grpc/pubsub/filterx/object-pubsub-message.cpp

```diff
@@ -0,0 +1,252 @@
+{
+  const gchar *value;
+  gsize length;
+  if (!filterx_object_extract_protobuf_ref(protobuf_object, &value, &length))
+    throw std::runtime_error("Argument is not a protobuf object");
+
+  if (!message.ParsePartialFromArray(value, length))
+    throw std::runtime_error("Failed to parse from protobuf object");
+}
+
+Message::Message(const Message &o, FilterXPubSubMessage *super_) : super(super_),
+  message(o.message)
+{
+}
+
+std::string
+Message::marshal(void)
+{
+  std::string serializedString = this->message.SerializePartialAsString();
+  return serializedString;
+}
+
+const google::pubsub::v1::PubsubMessage &
+Message::get_value() const
+{
+  return this->message;
+}
+
+// /* C Wrappers */
+
+static void
+_free(FilterXObject *s)
+{
+  FilterXPubSubMessage *self = (FilterXPubSubMessage *) s;
+
+  delete self->cpp;
+  self->cpp = NULL;
+
+  filterx_object_free_method(s);
+}
+
+static gboolean
+_truthy(FilterXObject *s)
+{
+  return TRUE;
+}
+
+static gboolean
+_marshal(FilterXObject *s, GString *repr, LogMessageValueType *t)
+{
+  FilterXPubSubMessage *self = (FilterXPubSubMessage *) s;
+
+  std::string serialized = self->cpp->marshal();
+
+  g_string_truncate(repr, 0);
+  g_string_append_len(repr, serialized.c_str(), serialized.length());
+  *t = LM_VT_PROTOBUF;
+  return TRUE;
+}
+
+static gboolean
+_map_to_json(FilterXObject *s, struct json_object **object, FilterXObject **assoc_object)
+{
+  FilterXPubSubMessage *self = (FilterXPubSubMessage *) s;
+
+  std::string serialized;
+  google::protobuf::util::MessageToJsonString(self->cpp->get_value(), &serialized);
+
+  *assoc_object =  filterx_json_new_from_repr(serialized.c_str(), serialized.length());
+  // *object = filterx_json_object_get_value(filterx_object_ref(*assoc_object));
+  *object = json_object_new_object();
+  json_object_object_foreach(filterx_json_object_get_value(*assoc_object), key, val)
+  {
+    json_object_object_add(*object, key, json_object_get(val));
+  }
+  return TRUE;
+}
+
+static void
+_init_instance(FilterXPubSubMessage *self)
+{
+  filterx_object_init_instance(&self->super, &FILTERX_TYPE_NAME(pubsub_message));
+}
+
+FilterXObject *
+_filterx_pubsub_message_clone(FilterXObject *s)
+{
+  FilterXPubSubMessage *self = (FilterXPubSubMessage *) s;
+
+  FilterXPubSubMessage *clone = g_new0(FilterXPubSubMessage, 1);
+  _init_instance(clone);
+
+  try
+    {
+      clone->cpp = new Message(*self->cpp, clone);
+    }
+  catch (const std::runtime_error &)
+    {
+      g_assert_not_reached();
+    }
+
+  return &clone->super;
+}
+
+
+gboolean
+_build_map(FilterXObject *key, FilterXObject *val, gpointer user_data)
+{
+  auto *attr = static_cast<std::map<std::string, std::string>*>(user_data);
+
+  GString *key_str = scratch_buffers_alloc();
+  GString *val_str = scratch_buffers_alloc();
+
+  if (!filterx_object_repr(key, key_str))
+    return FALSE;
+  if (!filterx_object_repr(val, val_str))
+    return FALSE;
```

Comment:
In the majority of the time these will already be strings, and `repr()` copies strings even. For strings we could call `get_value()`, to optimize out the copy, for the rest `repr()` is perfect.